### DISCUSSION TITLE
Stgit vs Git

### DISCUSSION THREAD

**Discussion by ckyoog** [stacked-git/stgit#366] [General]
Hi all,
I am trying to understand Stgit. After read the tutorial, I learned the workflow, the commands. But I still don't quite understand why you came up with this idea. Compared to git, what benefit that Sgit brings to users, what problems that only Stgit can resolve, or Stgit can resolve better? Appreciate if anybody can show me some examples. Thank you in advance!

**Comment by ctmarinas** [comment]
At the time I started this project, my aim was to have a lightweight tool on top of Git that closely resembled Quilt (patch management; at the time there wasn’t even a git send-email). These days Git grew lots of features that help manage patch series (e.g. git rebase) but StGit tends to be more convenient. The big feature that’s missing in Git (for my workflow) is the ability to remove patches from a series while still keeping them around to be re-added later. 

If you only commit changes and don’t need to rebase, reorder, rewrite commits, Git is fine. Even if you need all of these, Git is probably still sufficient. It’s only that StGit may make such workflows easier.

**Reply by ckyoog** [reply]
Thank you for the explanation. That's cool! I kinda get the idea of Stgit, and I think I will definitely try it, because for rebasing, reordering, rewriting the commits, they basically happen everyday during my work. I always love to adjust my commits before finally pushing them to the repo.

**Comment by aczapszys-apex** [comment]
> The big feature that’s missing in Git (for my workflow) is the ability to remove patches from a series while still keeping them around to be re-added later.

I've been trying to figure out how to do exactly that.  What does removing a patch from a series to save for latter look like?

**Reply by ctmarinas** [reply]
I have a set of git aliases to do this but they are not that great as they don't remember the order they've been pushed/popped, not that nice with per-branch patches. Basically popping a patch rebases the newer commits on top the popped commit parent and saves the popped commit into a tag. In the .gitconfig below, `series-stash` pops a patch from a series:

    [alias]
        current = !H=$(git symbolic-ref -q HEAD) && echo "${H#refs/heads/}" || echo "HEAD"
        get-base = !git config branch.$(git current).base || echo "origin/master"
        mkname = !git log -1 --pretty="%s" $1 | tr -s '[:blank:][:punct:]' '-' | tr '[:upper:]' '[:lower:]' "#"

        series = !git log --reverse --pretty=format:\"%C(auto)%h %s %Cblue[%an]%Creset\" $(git get-base)..$1
        series-all = !git series && git tag -l "stashed/*"
        series-apply = !git cherry-pick $1 && git tag -d $1 "#"
        series-delete = !git rebase --onto ${1:-HEAD}^ ${1:-HEAD} "#"
        series-edit = !git rebase -i --autosquash $(git base-or-merge)
        series-float = !git rebase --onto $1^ $1 && git cherry-pick $1 "#"
        series-format = !git format-patch --cover-letter ${1:+--subject-prefix=\"PATCH $1\"} -o patches-$(git current)${1:+-$1} $(git get-base).. "#"
        series-log = !git log --reverse $(git get-base)..$1
        series-rebase = !test -z \"$(git series HEAD --merges)\" && git rebase $(git base) --onto
        series-review = !git log --reverse --stat -p -C -M $(git get-base)..
        series-save = !git tag stashed/$(git current)/$(git mkname $1)
        series-stash = !git series-save $1 && git series-delete
        series-trailer = !git rebase -x \"git trailer \\\"$1\\\"\" $(git base-or-merge) "#"
        show-stashed = !git tag -l "stashed/*"

Of course, I still prefer stgit for any major series development but these days I tend to merge patches from others more than developing my own. As my workflow changed over the years (from developer to maintainer), what I'd like from stgit is the possibility to push/pop merge commits but I'm pretty lost in the new Rust codebase to attempt to do it ;) (not that I have much time either).

**Reply by jpgrayson** [reply]
> push/pop merge commits

Could you say more about that?

> I'm pretty lost in the new Rust codebase to attempt to do it

I feel sorry about that 😞. I didn't mean to make your creation inaccessible to you!

**Reply by ctmarinas** [reply]
> > push/pop merge commits

> Could you say more about that?

Like a patch being pushed/popped but instead of having a single parent, it is allowed to have two parents (or more; i.e. a merge commit). I normally merge patch series from others on specific topic branches, linear, and then mix all those topic branches together in a single main branch. I have this scripted and do a full merge of all branches every time but while testing sometimes it would be handy if I could just pop one of those branches out of the main branch, push it back etc.

Not essential really for StGit, I suspect my use-case is fairly limited (and can work around it already). Maybe that's a good incentive for me to learn Rust properly.

> > I'm pretty lost in the new Rust codebase to attempt to do it

> I feel sorry about that 😞. I didn't mean to make your creation inaccessible to you!

Ah, I am very grateful for your work and the Rust conversion. It's a lot faster and Rust makes it easier for you and more appealing to other contributors, I don't have any objections ;). Even if it was still in Python, I can't guarantee I would have contributed more.
